### ISSUE TITLE
Path of data directory

### ISSUE THREAD

**Issue by nheeren** [IndEcol/ODYM#5]
I corrected the path of the data directory again in commit 1821173194a04c1efb2e6c7d12f65a9804543a8b (line 313, 314). However, now I am a bit unsure if it was your intent to have the data directory *outside* of the model directory, @stefanpauliuk . In my understanding it should become part of it (i.e. be a sub-folder) eventually.

**Comment by stefanpauliuk** [comment]
With the new setup of the projects based on ODYM I propose two options after our discussion:
+ a path text file is present one level above the project folder, and that text file contains the links to ODYM (later to be replaced by ODYM being installed as package), to the data folder, and to the results folder.
+ If no such file is present, the project scripts looks for an ODYM, data, and results folder in the same directory as the project repository.

**Comment by stefanpauliuk** [comment]
Have created the following fancy text path file and the code below to parse it. Is that anywhere near a professional or at least durable workable solution to the path problem for the RECC project?

1) Path file:

*** Start of path file ***
*** Paths are for Windows machine ***
Path_ODYM = C:\Users\spauliuk\FILES\ARBEIT\PROJECTS\ODYM-RECC\ODYM_Model
Path_Data = C:\Users\spauliuk\FILES\ARBEIT\PROJECTS\ODYM-RECC\RECC_Database
Path_Results = C:\Users\spauliuk\FILES\ARBEIT\PROJECTS\ODYM-RECC\RECC_Results
*** End of path file ***


2) Path parsing:


def config_string_cutout(String, Code, leftstart, rightstart): # Returns substring between Code and EoL for given identifier in config file
# example: if Config_File_Line is "Data_Path_Network_1=K:\Research_Data" the function call
# config_string_cutout(Config_File_Line,'Data_Path_Network_1','=','\n') will return 'K:\Research_Data'
    Codeindex = String.find(Code,0,len(String))
    if Codeindex == -1:
        return 'None'
    else:
        Startindex = String.find(leftstart,Codeindex,len(String)) 
        Endindex   = String.find(rightstart,Codeindex,len(String))
        return String[Startindex +1:Endindex]


try:
    PathFile = open(os.path.abspath(os.path.join(os.path.join(os.path.dirname(__file__), '..'),'RECC_Paths.txt')), 'r') 
    PathText = PathFile.read()
    #Extract path names from main file
    Path_ODYM     = config_string_cutout(PathText,'Path_ODYM','=','\n').strip()
    Path_Data     = config_string_cutout(PathText,'Path_Data','=','\n').strip()
    Path_Results  = config_string_cutout(PathText,'Path_Results','=','\n').strip()
except: # use relativepaths in neighboring folders:
    
    Path_ODYM    = os.path.abspath(os.path.join(os.path.join(os.path.dirname(__file__), '..'),'ODYM_Model'))
    Path_Data    = os.path.abspath(os.path.join(os.path.join(os.path.dirname(__file__), '..'),'RECC_Database'))
    Path_Results = os.path.abspath(os.path.join(os.path.join(os.path.dirname(__file__), '..'),'RECC_Results'))

**Comment by nheeren** [comment]
Ideally I think something like this should happen: 

To me it seems like we should either use a plain python file or a YAML file: https://stackoverflow.com/q/5055042

I would suggest the following: 

1. As soon as ODYM is imported, it looks for the `odym_config.py` file in the same folder as the script is being executed (e.g. in the RECC folder).
2. If nothing is found it sets the path variables to `None` and expects the user to specify it manually e.g. by `data_path = "../RECC_Database"` 

The config.py would look something like this: 
```python
"""
This is the ODYM path file. ODYM will use the paths specified here to search for the necessary data and modules.
"""

odym_path = "C:\Users\spauliuk\FILES\ARBEIT\PROJECTS\ODYM-RECC\ODYM_Model"
data_path = "C:\Users\spauliuk\FILES\ARBEIT\PROJECTS\ODYM-RECC\RECC_Database"
results_path = "C:\Users\spauliuk\FILES\ARBEIT\PROJECTS\ODYM-RECC\RECC_Results"
```

Why do we need `Path_ODYM`?

**Comment by stefanpauliuk** [comment]
Hm. I don't quite understand your concerns. 
ODYM does not need to know paths, but the projects using ODYM, let's call one of them PROject.
PROject_Main.py will also be in a version-managed folder, and I wanted to place the path file one level above to conveniently avoid it being subject to version mgt:

--- ODYM: Somewhere on your machine
--- PROject Data: Somewhere on your machine

-- PROject Main folder
---Docu
---Paper
---Archive
--- ...
---PROject model code [git]
---PROject_Paths.py 

PROject_Paths.py containt the following code:

`
"""
This is the PROject path file. PROject_Main.py will use the paths specified here to search for the necessary data and modules.
"""

odym_path = "C:\Users\spauliuk\FILES\ARBEIT\PROJECTS\ODYM-RECC\ODYM_Model"
data_path = "C:\Users\spauliuk\FILES\ARBEIT\PROJECTS\ODYM-RECC\RECC_Database"
results_path = "C:\Users\spauliuk\FILES\ARBEIT\PROJECTS\ODYM-RECC\RECC_Results"
`

**Comment by nheeren** [comment]
True. It is the project that needs the path. Sorry for the confusion. 

Apparently, I still don't 100% understand your concern. Sorry, can you help me?

If excluding the config file from version management is your sole concern, then I would suggest to simply use a  a [`.gitignore` file](https://help.github.com/articles/ignoring-files/). This is the common way to exclude files from version management. So whenever you commit your code this file will be excluded by default. In our case we could have a routine that will create `paths.py` if it isn't found, e.g. when running ODYM in a new project folder for the first time.

If we placed `PROject_Paths.py` one directory above, it would end up in my `code` folder. This is a place where I would rather like to not have any files.

**Comment by nheeren** [comment]
P.S.: `.gitignore` can be directly included into the repo, as Konstantin does here: https://github.com/konstantinstadler/country_converter

**Comment by stefanpauliuk** [comment]
Seems like gitignore solves this problem!

Remaining is Konstantin's suggestion to create a package and install it locally. At least I will work on both ODYM and RECC simultaneously, and am OK with gitting both and importing ODYM via the path file.

**Comment by nheeren** [comment]
Ok. Looks like we have two todos for one of the next commits:

- [ ] include path file
- [ ] include a gitignore file and add the path to it

We could use [this .gitignore](https://github.com/github/gitignore/blob/master/Python.gitignore) file as a template. Should I give it a try next time?

**Comment by stefanpauliuk** [comment]
Just made a .gitignore for the RECC model folder, seems to work. If you create a RECC repo under Yale CIE I can push the things there and then we can finalize the "pathfinder".

**Comment by nheeren** [comment]
Here we go: https://github.com/YaleCIE/RECC-ODYM
You are admin and can change things, such as the repo name.
I selected the default pyhon gitignore file from github ;)

**Comment by stefanpauliuk** [comment]
My windows machine does not know what to do with the relative paths in sys.path. For the tutorials, I added both:

sys.path.insert(0, os.path.join('..', 'odym', 'modules')) # add ODYM module directory to system path, relative
sys.path.insert(0, os.path.join(os.getcwd(),'..', 'odym', 'modules')) # add ODYM module directory to system path, absolute

Will push, please check whether that works for you as well! If not, we can wrap the absolute path insertion in a try-except clause.

All 4 ODYM tutorials are now updated.

**Comment by nheeren** [comment]
Oh no! I was so hopeful that this would work for both OSes.  

What do you mean by "does not know what to do with the relative paths in sys.path"? What error is shown? I will install a virtual machine and test on Windows as well...

**Comment by stefanpauliuk** [comment]
For Windows: the paths "..\\ODYM .... etc " are added to sys.path, but when loading a module from that path the error "could not be found is thrown". So I added the absolute path as well.

**Comment by stefanpauliuk** [comment]
Loading paths from RECC_Paths.py is now working in the RECC main model script. Fille is under .gitignore and has the following content:

'''
This is the RECC model path file. RECC will use the paths specified here to search for the necessary data and modules.
'''

### Start of path file ###
### These paths are for Stefan Pauliuk's windows machine ###

odym_path = 'C:\\Users\\spauliuk\\FILES\\ARBEIT\\PROJECTS\\ODYM-RECC\\ODYM_Model' 
data_path = 'C:\\Users\\spauliuk\\FILES\\ARBEIT\\PROJECTS\\ODYM-RECC\\RECC_Database' 
results_path = 'C:\\Users\\spauliuk\\FILES\\ARBEIT\\PROJECTS\\ODYM-RECC\\RECC_Results' 

### End of path file ###

In the script the paths are used with: 

import RECC_Paths

RECC_Paths.odym_path ....

**Comment by nheeren** [comment]
> For Windows: the paths "..\ODYM .... etc " are added to sys.path, but when loading a module from that path the error "could not be found is thrown". So I added the absolute path as well.

I am not sure if this issue is still up to date. I installed everything on Windows and it seemed to work in Tutorial 1.  There it says: 

```python
sys.path.insert(0, os.path.join('..', 'odym', 'modules'))
sys.path.insert(0, os.path.join(os.getcwd(),'..', 'odym', 'modules')) 
```
So are we good?

**Comment by stefanpauliuk** [comment]
Yes! the workaround/solution I propose is to add both: absolute paths with getcwd and relative paths with ..
Please close the issue if that works with you!
